File: 332. Reconstruct Itinerary.py
from collections import defaultdict, deque
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        '''
        build graph and use dfs to find path
        '''
        def build_graph(tickets):
            for t in tickets:
                G[t[0]].append(t[1])
            for t in G:
                G[t] = deque(sorted(G[t], reverse=True))

        def find_route(S):
            route.append(S)
            if len(route) == len(tickets) +1:
                return True
            for i in range(len(G[S])):
                A = G[S].pop()
                if find_route(A):
                    return True
                G[S].appendleft(A)
            route.pop()
            return False
        
        G = defaultdict(list)
        build_graph(tickets)
        route = []
        find_route('JFK')
        return route

File: 332. Reconstruct Itinerary.java
/**
 * greedy DFS & backtracking
 * point is using the return value of dfs to mark if found
 * Time: build graph O(V + ElogE) dfs+backtrack O(E)
 * Space: graph O(E) stack O(E) worst
 */
class Solution {
    List<String> ans;
    public List<String> findItinerary(List<List<String>> tickets) {
        HashMap<String, List<String>> graph = new HashMap<>();
        int edges = tickets.size();
        for (List<String> t : tickets) {
            if (!graph.containsKey(t.get(0))) {
                List<String> temp = new LinkedList<>();
                temp.add(t.get(1));
                graph.put(t.get(0), temp);
            } else {
                graph.get(t.get(0)).add(t.get(1));
            }
        }
        for (String key : graph.keySet()) {
            Collections.sort(graph.get(key));
        }
        backtrack("JFK", graph, new ArrayList<>(), edges);
        return ans;
    }
    private boolean backtrack(String curr, HashMap<String, List<String>> graph, List<String> ls, int edges) {
        ls.add(curr);
        if (ls.size() == edges+1) {
            ans = new ArrayList<>(ls);
            return true;
        }
        int size = 0;
        if (graph.containsKey(curr)) {
            size = graph.get(curr).size();
        }
        for (int i=0; i<size; i++) {
            String next = graph.get(curr).remove(0);
            if (backtrack(next, graph, ls, edges)) {
                return true;
            }
            graph.get(curr).add(next);
        }
        ls.remove(ls.size()-1);
        return false;
    }
}